### ISSUE TITLE
Bug in prediction code

### ISSUE THREAD

**Issue by malkin1729** [vered1986/self_talk#1]
I believe there is a bug around the following line in the baseline prediction code:
https://github.com/vered1986/self_talk/blob/bbfc675b61a582aa00047ca31334f8fa75efcbad/source/predictors/multiple_choice_baseline_predictor_unsupervised.py#L224
Specifically, a batch given as input to the LM is constructed by stacking several candidate lists of tokens, which are first made the same length by padding with the token 0. However, the scoring function that follows (L252-267) takes the likelihoods of these padding tokens into account when computing total log-likelihood of each candidate.

A probe that shows this is indeed a bug: change the default padding token in that line from 0 to 50256. The predictions and accuracies obtained by running `bash experiments/piqa/predict_baseline_zero_shot.sh 0 dev` will change.

If the code is modified to ignore the padding tokens when scoring, the baseline results improve significantly. For example, on the PIQA validation set, the accuracy of GPT-XL, computed correctly, is 69.6% (compared with 62.6% as reported in Table 2 of your paper).

**Comment by vered1986** [comment]
Thanks for letting me know. I will look at it early next week. 

Both the model and the baseline scripts default to padding of zero when there is no padding token, so I'll re-run everything after changing the cross entropy loss to ignore the padding tokens.

**Comment by malkin1729** [comment]
Thanks. I'm curious to see what the results will be like with the change.

This is really interesting work and a helpful codebase. Stumbled upon it while working on a project that is also about toying with contexts for better pretrained LM inference.

**Comment by vered1986** [comment]
Thanks again for pointing out the bug. I changed the padding to:

```
tokenizer.pad_token = tokenizer.eos_token
```

When `tokenizer.pad_token` is `None`. I also changed the cross-entropy loss to ignore the padded tokens. 

The (development) accuracy is indeed higher in general, but the trend generally holds (baseline < knowledge model ~ self-talk). 

<table border="1" class="dataframe">
  <thead>
    <tr>
      <th>Task</th>
      <th>Type</th>
      <th>LM</th>
      <th>Knowledge Source</th>
      <th>Accuracy</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td rowspan="3">COPA</td>
      <td>Baseline</td>
      <td>GPT2-L</td>
      <td>-</td>
      <td>68.0</td>
    </tr>
    <tr>
      <td>Ext. Knowledge</td>
      <td>GPT2-XL</td>
      <td>COMET</td>
      <td>67.0</td>
    </tr>
    <tr>
      <td>Self-Talk</td>
      <td>GPT2-XL</td>
      <td>DistilGPT2</td>
      <td><b>69.0</b></td>
    </tr>
    <tr>
      <td rowspan="3">CommonsenseQA</td>
      <td>Baseline</td>
      <td>GPT2-XL</td>
      <td>-</td>
      <td>36.7</td>
    </tr>
    <tr>
      <td>Ext. Knowledge</td>
      <td>GPT2-L</td>
      <td>Google Ngrams</td>
      <td><b>42.6</b></td>
    </tr>
    <tr>
      <td>Self-Talk</td>
      <td>GPT2-XL</td>
      <td>XLNet-B</td>
      <td>41.7</td>
    </tr>
    <tr>
      <td rowspan="3">MC-TACO</td>
      <td>Baseline</td>
      <td>GPT2-XL</td>
      <td>-</td>
      <td>61.2</td>
    </tr>
    <tr>
      <td>Ext. Knowledge</td>
      <td>GPT2-XL</td>
      <td>COMET</td>
      <td><b>63.2</b></td>
    </tr>
    <tr>
      <td>Self-Talk</td>
      <td>GPT2-XL</td>
      <td>GPT</td>
      <td>63.0</td>
    </tr>
    <tr>
      <td rowspan="3">Social IQa</td>
      <td>Baseline</td>
      <td>GPT2-L</td>
      <td>-</td>
      <td>43.7</td>
    </tr>
    <tr>
      <td>Ext. Knowledge</td>
      <td>GPT2-XL</td>
      <td>ConceptNet</td>
      <td><b>46.3</b></td>
    </tr>
    <tr>
      <td>Self-Talk</td>
      <td>GPT2-XL</td>
      <td>XLNet-B</td>
      <td>45.9</td>
    </tr>
    <tr>
      <td rowspan="3">PIQA</td>
      <td>Baseline</td>
      <td>GPT2-XL</td>
      <td>-</td>
      <td>70.2</td>
    </tr>
    <tr>
      <td>Ext. Knowledge</td>
      <td>GPT2-XL</td>
      <td>ConceptNet</td>
      <td>70.9</td>
    </tr>
    <tr>
      <td>Self-Talk</td>
      <td>GPT2-XL</td>
      <td>GPT2</td>
      <td><b>71.0</b></td>
    </tr>
    <tr>
      <td rowspan="3">WinoGrande</td>
      <td>Baseline</td>
      <td>GPT2-XL</td>
      <td>-</td>
      <td><b>55.6</b></td>
    </tr>
    <tr>
      <td>Ext. Knowledge</td>
      <td>GPT2-XL</td>
      <td>COMET</td>
      <td>55.2</td>
    </tr>
    <tr>
      <td>Self-Talk</td>
      <td>GPT2-XL</td>
      <td>DistilGPT2</td>
      <td>55.2</td>
    </tr>
</tbody>
</table>

**Comment by malkin1729** [comment]
Thanks for the quick fix and for sharing the new numbers!

**Comment by malkin1729** [comment]
Sharing the results of the aforementioned project (ACL'22, to appear).
https://arxiv.org/abs/2110.08294
It would be interesting to see if there is an additive effect of coherence boosting on self-talk, with contrastive scoring to increase the effect of the self-talk explanations.

**Comment by vered1986** [comment]
Nice, thanks for sharing your paper! It might work well when using multiple (instead of one) clarifications.

**Comment by malkin1729** [comment]
I just ran a quick experiment with CommonsenseQA, using a score 
(1-alpha)*(score with clarifications)+alpha*(score without clarifications)
with alpha<0 to boost the self-talk effect. There indeed seems to be a benefit, with peak accuracy improving an additional 2% over self-talk alone (43.8%).

Because contrasting with unconditional accuracy (as done in our paper) gives an improvement of over 10% on this dataset, we could think of selecting **three** log-linear mixture weights instead of two: for the standard query+response, for the unconditional response, and for the response with clarifications. It should be very quick to test out (forthcoming!).

If you'd like to discuss further, perhaps we could take it to email?

**Comment by vered1986** [comment]
Sure, I'd love to hear more, we can discuss it in email. It would probably take me a while to find time read the paper in depth and respond :)
